### ISSUE TITLE
[M12] Dedupe allowed_ids before the rng.choice fallback in constrained decode

### ISSUE THREAD

**Issue by travisgalloway** [travisgalloway/monica#285]
Follow-up from PR #284 (issue #226, completion-list logit masking).

Copilot's sixth review round on #284 surfaced one remaining low-confidence item that was not addressed, because `/automerge` had reached its 5-iteration safety cap. Recording it here rather than losing it.

**Item:** `src/serve/sampling.py:83` — `allowed_ids` is not deduplicated before the `rng.choice` fallback path. If the list contains duplicates, the fallback biases selection toward the duplicated tokens.

**Severity:** cosmetic / edge-case. The constrained-decode contract is *not* broken — every sampled token is still drawn from `allowed_ids`. Only the distribution over allowed tokens is skewed, and only on the fallback path.

**Fix sketch:** dedupe (preserving order) before the fallback selection, plus a regression test asserting a uniform-ish draw over a duplicate-bearing `allowed_ids`.

For context, PR #284 went through five remediation cycles; one of them fixed a genuine correctness bug (`78e842d`: the all-non-finite fallback could return a token outside `allowed_ids` when repetition control banned everything). This remaining item is materially smaller than that one.

## Capability
A person sampling with `allowed_ids` gets a genuinely uniform draw from the all-non-finite
fallback, rather than one biased toward ids the caller happened to list more than once.

## Definition of done
- [x] `allowed_ids` is deduplicated, **preserving first-seen order**, before the fallback
      `rng.choice(ids)` at `src/serve/sampling.py:127`.
- [x] Dedup happens once, at the point `ids` is built (`src/serve/sampling.py:81`), so the
      in-range filter, the empty-set guard and the fallback all see the same set.
- [x] The existing guards still fire on the same inputs: empty `allowed_ids` raises, and an
      `allowed_ids` whose every entry is out of `[0, vocab)` raises with the existing message.
- [x] A regression test asserts a uniform-ish draw over a duplicate-bearing `allowed_ids` under a
      fixed seed, and that the returned id is always within the set.
- [x] The normal (finite-logits) path is unchanged — no behavioural difference when logits are
      usable, which is the overwhelmingly common case.
- [x] `tests/test_constrained_sampling.py` and `tests/test_repetition_penalty.py` still pass.

## Edge cases
- Every id duplicated (e.g. `[5, 5, 5]`) → one candidate, drawn deterministically.
- Duplicates that are also out of range → dropped by the range filter, and the resulting empty set
  raises rather than silently narrowing.
- A single-element `allowed_ids` → unchanged behaviour.
- Very large `allowed_ids` (whole-vocab masks) → dedup must not become the hot path's cost.

## Verification
- `.venv/bin/python -m pytest tests/test_constrained_sampling.py tests/test_repetition_penalty.py -q`
- New case in `tests/test_constrained_sampling.py` covering the duplicate-bearing fallback.

**Comment by travisgalloway** [comment]
Fixed in #310 (`0ef5f81`). Dedupe lands once at the `ids` build site in `src/serve/sampling.py`, preserving first-seen order, so the range filter, the empty-set guard, the mask and the fallback draw all read the same set. New bias test fails on `main` (id 7 drawn 0.760 of 2000 draws) and passes on the branch. Full suite 1645 passed / 12 env skips; smoke gate green.
